tests/test_internals/test_roi_extractors_module.py

```diff
@@ -143,6 +144,92 @@ def test_device_object_and_metadata_mix(self):
+    def setUp(self):
+        self.session_start_time = datetime.now().astimezone()
+        self.nwbfile = NWBFile(
+            session_description="session_description",
+            identifier="file_id",
+            session_start_time=self.session_start_time,
+        )
+
+        self.metadata = dict(Ophys=dict())
+
+        self.device_name = "optical_device"
+        self.device_metadata = dict(name=self.device_name)
+        self.metadata["Ophys"].update(Device=[self.device_metadata])
+
+        self.optical_channel_metadata = dict(
+            name="optical_channel",
+            emission_lambda=np.nan,
+            description="description",
+        )
+
+        self.imaging_plane_name = "imaging_plane_name"
+        self.imaging_plane_description = "imaging_plane_description"
+        self.imaging_plane_metadata = dict(
+            name=self.imaging_plane_name,
+            optical_channel=[self.optical_channel_metadata],
+            description=self.imaging_plane_description,
+            device=self.device_name,
+            excitation_lambda=np.nan,
+            indicator="unknown",
+            location="unknown",
+        )
+
+        self.metadata["Ophys"].update(ImagingPlane=[self.imaging_plane_metadata])
+
+    def test_add_imaging_plane(self):
+
+        add_imaging_plane(nwbfile=self.nwbfile, metadata=self.metadata)
+
+        imaging_planes = self.nwbfile.imaging_planes
+        assert len(imaging_planes) == 1
+        assert self.imaging_plane_name in imaging_planes
+
+        imaging_plane = imaging_planes[self.imaging_plane_name]
+        assert imaging_plane.description == self.imaging_plane_description
+
+    def test_not_overwriting_imaging_plane_if_same_name(self):
+
+        add_imaging_plane(nwbfile=self.nwbfile, metadata=self.metadata)
+
+        self.imaging_plane_metadata["description"] = "modified description"
+        add_imaging_plane(nwbfile=self.nwbfile, metadata=self.metadata)
+
+        imaging_planes = self.nwbfile.imaging_planes
+        assert len(imaging_planes) == 1
+        assert self.imaging_plane_name in imaging_planes
+
+    def test_add_two_imaging_planes(self):
+
+        # Add the first imaging plane
+        first_imaging_plane_name = "first_imaging_plane_name"
+        first_imaging_plane_description = "first_imaging_plane_description"
+        self.imaging_plane_metadata["name"] = first_imaging_plane_name
+        self.imaging_plane_metadata["description"] = first_imaging_plane_description
+        add_imaging_plane(nwbfile=self.nwbfile, metadata=self.metadata)
+
+        # Add the second imaging plane
+        second_imaging_plane_name = "second_imaging_plane_name"
+        second_imaging_plane_description = "second_imaging_plane_description"
+        self.imaging_plane_metadata["name"] = second_imaging_plane_name
+        self.imaging_plane_metadata["description"] = second_imaging_plane_description
+        add_imaging_plane(nwbfile=self.nwbfile, metadata=self.metadata)
```

Comment:
Yes, I think this is confusing and I think it stems from a design decision that was made about our metadata format. In order to mimic the nwb schema the metadata for imagine plane is stored at the top level of ophys as a list. The reason for this, as far as I understand, is that the `imaging_plane` in the nwb-schema is a group of objects.  Each of the imaging plane in this group are then stored as links in all the objects that require them (e.g. TwoPhotonSeries). I think that the decision was make to make our own schema in nwb-conversion tools mimic as much the true schema. 

The important outcome is that the metadata for the `imaging_plane` in our metadata dict is a list and therefore an specification of which element/index of the list do we want to add becomes necessary. That's the role of `imagine_plane_index`. If you see the previous code that this is removing it was assumed that the element was always the first which I think works right now because we never had a conversion with more than one plane. I changed this to make this assumption explicit and make it a parameter that can be changed if it is required. 

Maybe there is a better name? as we don't have examples of conversion data with more than two imaging planes I don't know what's the best way to handle this.

Reply:
I see, thanks for clarifying it.